flow/connectors/mongo/mongo.go

```diff
@@ -268,100 +267,125 @@ func (c *MongoConnector) PullRecords(
-			items.AddColumn(DefaultFullDocumentColumnName, types.QValueJSON{Val: "{}"})
-		}
 
-		if operationType, ok := changeDoc["operationType"]; ok {
-			switch operationType {
-			case "insert":
-				if err := addRecord(ctx, &model.InsertRecord[model.RecordItems]{
-					BaseRecord:           model.BaseRecord{CommitTimeNano: clusterTimeNanos},
-					Items:                items,
-					SourceTableName:      sourceTableName,
-					DestinationTableName: destinationTableName,
-				}); err != nil {
-					return fmt.Errorf("failed to add insert record: %w", err)
-				}
-			case "update", "replace":
-				if err := addRecord(ctx, &model.UpdateRecord[model.RecordItems]{
-					BaseRecord:           model.BaseRecord{CommitTimeNano: clusterTimeNanos},
-					NewItems:             items,
-					SourceTableName:      sourceTableName,
-					DestinationTableName: destinationTableName,
-				}); err != nil {
-					return fmt.Errorf("failed to add update record: %w", err)
+			if operationType, ok := changeDoc["operationType"]; ok {
+				switch operationType {
+				case "insert":
+					if err := addRecord(ctx, &model.InsertRecord[model.RecordItems]{
+						BaseRecord:           model.BaseRecord{CommitTimeNano: clusterTimeNanos},
+						Items:                items,
+						SourceTableName:      sourceTableName,
+						DestinationTableName: destinationTableName,
+					}); err != nil {
+						return fmt.Errorf("failed to add insert record: %w", err)
+					}
+				case "update", "replace":
+					if err := addRecord(ctx, &model.UpdateRecord[model.RecordItems]{
+						BaseRecord:           model.BaseRecord{CommitTimeNano: clusterTimeNanos},
+						NewItems:             items,
+						SourceTableName:      sourceTableName,
+						DestinationTableName: destinationTableName,
+					}); err != nil {
+						return fmt.Errorf("failed to add update record: %w", err)
+					}
+				case "delete":
+					if err := addRecord(ctx, &model.DeleteRecord[model.RecordItems]{
+						BaseRecord:           model.BaseRecord{CommitTimeNano: clusterTimeNanos},
+						Items:                items,
+						SourceTableName:      sourceTableName,
+						DestinationTableName: destinationTableName,
+					}); err != nil {
+						return fmt.Errorf("failed to add delete record: %w", err)
+					}
+				case "invalidate":
+					return errors.New("change stream invalidated, please resync the mirror")
+				default:
+					return fmt.Errorf("unsupported operationType: %s", operationType)
 				}
-			case "delete":
-				if err := addRecord(ctx, &model.DeleteRecord[model.RecordItems]{
-					BaseRecord:           model.BaseRecord{CommitTimeNano: clusterTimeNanos},
-					Items:                items,
-					SourceTableName:      sourceTableName,
-					DestinationTableName: destinationTableName,
-				}); err != nil {
-					return fmt.Errorf("failed to add delete record: %w", err)
+			}
+		}
+		// this branch will almost always be hit with the .Next() API
+		if err := changeStream.Err(); err != nil {
```

Comment:
This makes me wonder if we should even attempt to prolong the lifetime of PullRecord when we get a context deadline exceeded, and instead when idleTimeout is reached we just exit and let the next PullRecord  continue from where this left off.

Reply:
need to avoid creating empty batches if you do that

Reply:
cc @serprex  -- i know we do this in mysql, what's the benefit of prolonging an existing activity (comparing to starting a new activity)

Reply:
discussed offline with @serprex -- it will lead to cluttered UI and wasted effort processing empty batches. and since postgres require connection to stay open, it would be better to handle all connectors consistently.

I'm wondering if the best option here is to just call changestream.Next() without an idleTimeout deadline -- but in a separate goroutine where we pass decoded events back via a channel. This will allow us to prolong Activity lifetime in the main loop, something like:
```
select {
   case <- ctx.done():
          break
   case <-time.After(config.idleTimeout):
         saveResumeToken()
   case event  <- eventCh:
         process event or error
   case        
```

Reply:
That'd work, so long as:
1. this cannot be done once a record is received, so `saveResumeToken` here must not race with receiving a change which updates resume token
2. `saveResumeToken` must be threadsafe

Other option is to have a `time.Ticker` that's used in main loop, but you want to stop the ticker once messages are being received

In MySQL I went the context timeout route since there's symmetry where before first record we're waiting on hour timer & after first record we're waiting on sync interval timer

Reply:
good catch. it looks like changestream API isn't thread-safe when it comes to reading/write the `cs.resumeToken` field. 

I think this means we are left with options of:
- create a new changestream instance when context times out
- use TryNext instead of Next which is non-blocking 

I think second approach means we need to keep polling on TryNext, which i am not a fan.

Reply:
Discussed with @heavycrystal  -- merged change linked above.